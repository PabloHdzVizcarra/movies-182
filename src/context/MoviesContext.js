import React, { createContext, useReducer, useCallback, useState } from 'react';
import { moviesReducer } from '../reducers/moviesReducer';
import { types } from '../types/types';
import { db, firebase } from '../libs/firebase';

export const MoviesContext = createContext();



export const MoviesProvider = ({ children }) => {

  const [state, dispatch] = useReducer(moviesReducer, {
    movies: [],
    favoriteMovies: [],
    searchMovie: [],
    movie: {},
    loadingMovies: true
  });

  const [errorForm, setErrorForm] = useState({
    error: false,
    message: ''
  });
  
  const { movies, favoriteMovies, searchMovie, movie, loadingMovies } = state;
  

  const getTopRatedMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/top_rated?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingTopRatedMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getPopularMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/popular?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingPopularMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getUpcomingMovies = useCallback(async () => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/upcoming?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingUpcomingMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getMoviesFirebase = async(uid) => {

    try {
      
      let allFavoritesMovies = [];

      await db.collection(uid)
      .doc("movies")
      .collection("favorites")
      .get()
      .then((querySnapshot) => {
        querySnapshot.forEach((doc) => {
          let movie;
          movie = { ...doc.data() }
          movie.docID = doc.id;
          allFavoritesMovies.push(movie)
        });
      });

      // const reversedArray = allFavoritesMovies.reverse();

      dispatch({
        type: types.getFavoritesMoviesFirebase,
        payload: allFavoritesMovies
      })

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieSearch = async (key) => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/search/movie?api_key=${apiKey}&query=${key}`);
      
      const { results } = await data.json();
      console.log(results);

      if (results.length === 0) {
        return setErrorForm({
          error: true,
          message: 'Intenta buscando otra pelicula'
        })
      }

      setErrorForm({
        error: false,
        message: ''
      })

      dispatch({
        type: types.searchMovies,
        payload: results
      });

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieID = async (movieID) => {

    try {

      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/movie/${movieID}?api_key=${apiKey}&language=es-ES`);

      const resp = await data.json();
      
      dispatch({
        type: types.selectedMovie,
        payload: resp
      })
      
    } catch (error) {
      console.log(error);
    }
  }



  const addFavoriteMovie = async (movie) => {

    const {
      poster_path,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    } = movie;

    const imgMovie = `https://image.tmdb.org/t/p/w500/${poster_path}`;

    const favoriteMovie = {
      imgMovie,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    }

    const { uid } = firebase.auth().currentUser;

    if (uid) {

      const { id } = await db.collection(uid).doc('movies').collection('favorites').add(favoriteMovie);

      favoriteMovie.firebaseID = id;

      dispatch({
        type: types.addFavoritesMovies,
        payload: favoriteMovie
      })
      
      console.log('PELICULA AGREGADA A FAVORITOS');

    } else {
      throw new Error('No hay usuario autenticado')
    }

  }

  const deleteMovieFromFavorites = async (userID, docID) => {
    console.log(docID);

    dispatch({
      type: types.removeMovieFromFavorites,
      payload: docID
    })
    

    await db.collection(userID).doc('movies').collection('favorites').doc(docID)
      .delete();

    return console.log(`PELICULA ELIMINADA`);
  }
  
  return (
    <MoviesContext.Provider value={{
      movie,
      movies,
      favoriteMovies,
      searchMovie,
      loadingMovies,
      errorForm,
      getMovieID,
      getPopularMovies,
      getTopRatedMovies,
      getUpcomingMovies,
      getMovieSearch,
      addFavoriteMovie,
      getMoviesFirebase,
      deleteMovieFromFavorites,
      dispatch
    }}>
      {children}
    </MoviesContext.Provider>
  )
}