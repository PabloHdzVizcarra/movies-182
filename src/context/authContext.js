import React from 'react'
import { types } from '../types/types';

const AuthStateContext = React.createContext();
const AuthDispatchContext = React.createContext();

function authReducer(state, action) {
  switch (action.type) {

    case types.setActiveUser:
      return {
        ...state,
        activeUser: action.payload,
        isActived: true
      }

    case types.logoutUser:
      return {
        ...state,
        activeUser: {},
        isActived: false
      }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}


function AuthProvider({ children }) {
  const [state, dispatch] = React.useReducer(authReducer, {
    activeUser: {},
    isActived: false
  });

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  )
}

function useAuthState() {
  const context = React.useContext(AuthStateContext);

  if (context === undefined) {
    throw new Error('useAuthState must be used within AuthProvider')
  }

  return context
}

function useAuthDispatch() {
  const context = React.useContext(AuthDispatchContext);

  if (context === undefined) {
    throw new Error('useAuthDispatch must be used within AuthProvider')
  }

  return context
}

export {
  AuthProvider,
  useAuthState,
  useAuthDispatch
}