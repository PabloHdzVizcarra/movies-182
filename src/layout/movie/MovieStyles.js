import styled from "styled-components";
import { respondTo } from "../../styles/_respondTo";

export const ContainerMovie = styled.div`
  display: grid;
  background-color: #ffffff;
  grid-template-columns: auto 1fr;

  ${respondTo.xs`

  `}
  ${respondTo.md`
    padding-bottom: 20px;
  `}
`;

export const TitleContainer = styled.div`
  grid-column: 1/3;
  padding: 10px 0 10px 20px;

  h2 {
    margin-top: 5px;
    margin-bottom: 5px;
    color: #ff6d00;
  }

  p {
    margin-top: 5px;
    margin-bottom: 5px;
  }

  ${respondTo.xs`
    h2 {
      font-size: 1.8rem;
    }
  `}
`;

export const ImageContainer = styled.div`
  padding: 0 20px 10px 20px;
  grid-column: 1/3;

  

  img {
    width: 100%;
    -webkit-box-shadow: 10px 9px 10px -5px rgba(0,0,0,0.75);
    -moz-box-shadow: 10px 9px 10px -5px rgba(0,0,0,0.75);
    box-shadow: 10px 9px 10px -5px rgba(0,0,0,0.75);
  }

  ${respondTo.xs`
    display: grid;
    justify-content: center;
    padding: 0 25px 0 25px;

    img {
    }
  `}

  ${respondTo.sm`
    justify-content: center;
    img {
      width: 100%;
    }
  `}

  ${respondTo.md`
    display: grid;
    grid-column: 1/2;
    padding-right: 0px;

    h2 {
      font-size: 3rem;
      color: #f9a825;
    }

    img {
      border-radius: 6px 0 0 6px;
    }
  `}
`;

export const StyleContainer = styled.div`
  margin: 10px;
  padding: 20px;
  border-radius: 6px;
  background-color: rgba(82, 72, 156, .1);
  grid-column: 1/3;
    
  p {
    margin-top: 5px;
    margin-bottom: 5px;
  }
  
  p:nth-child(1) {
    margin-bottom: 10px;
  }

  span {
    font-weight: bold
  }

  ${respondTo.sm`
    margin: 35px;
    justify-content: center;
    grid-column: 1/2;

    img {
      width: 100%;
    }
  `}
  ${respondTo.md`
    grid-column: 2/3;
    margin: 0px 15px 0 0;
    border-radius: 0px 6px 6px 0;

  `}

  ${respondTo.lg`

    font-size: 1.1rem;
  `}
  
`;

export const IconHeart = styled.div`
  display: flex;
  align-items: end;

  svg {
    font-size: 2rem;
    color: #f34747;
    cursor: pointer;
    margin-left: 20px;

    &:hover {
      color: #f12727;
    }

    &:focus {
      font-size: 2.2rem;
    }
  }
`;
