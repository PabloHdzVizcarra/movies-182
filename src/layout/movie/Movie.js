import React, { useContext, useEffect } from "react";
import { useParams } from "react-router-dom";
import { MoviesContext } from "../../context/MoviesContext";
import { types } from "../../types/types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  ContainerMovie,
  ImageContainer,
  StyleContainer,
  IconHeart,
  TitleContainer,
} from "./MovieStyles";
import moment from "moment";
import "moment/locale/es";
import { useAuthState } from "../../context/authContext";
import { NotFound } from "../../pages/not-found/NotFound";
import { Spinner } from "../../components/spinner/Spinner";
moment.locale("es");

export const Movie = () => {
  const user = useAuthState();

  const {
    movie,
    getMovieID,
    addFavoriteMovie,
    dispatch,
    getMoviesFirebase,
    searchMovieFavorites,
  } = useContext(MoviesContext);
  const { movieID } = useParams();

  useEffect(() => {
    let mounted = true;
    if (mounted) {
      getMovieID(movieID);
    }

    if (user.activeUser.uid) {
      getMoviesFirebase(user.activeUser.uid);
    }

    return () => {
      mounted = false;
      dispatch({
        type: types.removeMovie,
      });
    };
    //eslint-disable-next-line
  }, [user.activeUser.uid]);

  if (Object.keys(movie).length === 0) {
    return <Spinner />;
  }

  if (!movie.imdb_id) {
    return <NotFound />;
  }

  const handleAddFavorite = () => {
    addFavoriteMovie(movie);
  };

  const {
    genres,
    id,
    overview,
    production_countries,
    release_date,
    tagline,
    popularity,
    vote_average,
    vote_count,
    poster_path,
    title,
    original_title,
  } = movie;

  const newDate = moment(release_date).format("DD MMMM YYYY");

  const allGeneres = genres.map((genre) => genre.name).join("/");
  const allCountrys = production_countries
    .map((genre) => genre.name)
    .join(" | ");

  return (
    <ContainerMovie className="animate__animated animate__fadeIn">
      <TitleContainer>
        <h2>{title}</h2>
        <p>{tagline}</p>
      </TitleContainer>

      <ImageContainer>
        <img
          src={
            poster_path
              ? `https://image.tmdb.org/t/p/w500/${poster_path}`
              : "https://images.unsplash.com/photo-1518676590629-3dcbd9c5a5c9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=80"
          }
          alt="No hay Imagen Disponible"
        />
      </ImageContainer>

      <StyleContainer>
        <div
          style={{
            display: "flex",
            flexDirection: "column",
            height: "100%",
            justifyContent: "center",
          }}
        >
          <p>{overview}</p>
          <p>
            <span>Titulo Original:</span> {original_title}
          </p>
          <p>
            <span>Genero:</span> {allGeneres}
          </p>
          <p>
            <span>Pais de produccion:</span> {allCountrys}
          </p>
          <p>
            <span>Lanzamiento:</span> {newDate}
          </p>
          <p>
            <span>Likes:</span> {popularity}
          </p>
          <p>
            <span>Calificacion:</span> {vote_average}
          </p>
          <p>
            <span>Numero de votos:</span> {vote_count}
          </p>
          {user.isActived && !searchMovieFavorites(id) && (
            <IconHeart>
              <p style={{ color: "#ff6d00", fontWeight: "bold" }}>
                {!searchMovieFavorites(id)
                  ? "AÃ±adir a favoritos"
                  : "Pelicula guardada en favoritos"}
              </p>
              <FontAwesomeIcon icon={"heart"} onClick={handleAddFavorite} />
            </IconHeart>
          )}
        </div>
      </StyleContainer>
    </ContainerMovie>
  );
};
