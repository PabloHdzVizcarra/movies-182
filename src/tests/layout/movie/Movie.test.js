import React from "react";
import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { Movie } from "../../../layout/movie/Movie";
import { AuthProvider, AuthStateContext } from "../../../context/authContext";
import { MoviesContext } from "../../../context/MoviesContext";
import { MemoryRouter } from "react-router-dom";

describe("Test in <Movie />", () => {
  jest.mock("react-router-dom", () => ({
    useParams: () => ({
      movieID: 612706,
    }),
  }));

  test("the film must be shown correctly", () => {
    const mockMoviesContext = {
      movie: {
        id: 612706,
        title: "Work It: Al ritmo de los sueños",
        genres: [
          {
            name: "Comedia",
          },
        ],
        production_countries: [
          {
            name: "Mexico",
          },
        ],
      },
      getMovieID: jest.fn(),
    };

    const wrapper = mount(
      <AuthProvider>
        <MoviesContext.Provider value={mockMoviesContext}>
          <MemoryRouter>
            <Movie />
          </MemoryRouter>
        </MoviesContext.Provider>
      </AuthProvider>
    );

    const textNode = wrapper.find("h2").text().trim();
    expect(textNode).toBe(mockMoviesContext.movie.title);
  });

  test("if active user must show the button to add movie to favorites", () => {
    const mockMoviesContext = {
      movie: {
        id: 612706,
        title: "Work It: Al ritmo de los sueños",
        genres: [
          {
            name: "Comedia",
          },
        ],
        production_countries: [
          {
            name: "Mexico",
          },
        ],
      },
      getMovieID: jest.fn(),
      searchMovieFavorites: jest.fn(),
      getMoviesFirebase: jest.fn(),
    };

    const wrapper = mount(
      <AuthProvider>
        <AuthStateContext.Provider
          value={{
            isActived: true,
            activeUser: {
              uid: "15asdsa",
            },
          }}
        >
          <MoviesContext.Provider value={mockMoviesContext}>
            <MemoryRouter>
              <Movie />
            </MemoryRouter>
          </MoviesContext.Provider>
        </AuthStateContext.Provider>
      </AuthProvider>
    );

    const textNode = wrapper.find("p").getElements()[9].props.children;
    expect(textNode).toBe("Añadir a favoritos");
  });
});
