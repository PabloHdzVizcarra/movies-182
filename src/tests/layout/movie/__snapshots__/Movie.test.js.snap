// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Movie /> if active user must show the button to add movie to favorites 1`] = `
<div
  className="sc-AxjAm kpQkWH animate__animated animate__fadeIn"
>
  <div
    className="sc-AxirZ cTYuej"
  >
    <h2>
      Work It: Al ritmo de los sueños
    </h2>
    <p />
  </div>
  <div
    className="sc-AxiKw dqhJqz"
  >
    <img
      alt="No hay Imagen Disponible"
      src="https://images.unsplash.com/photo-1518676590629-3dcbd9c5a5c9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=80"
    />
  </div>
  <div
    className="sc-AxhCb fYZkjQ"
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <p />
      <p>
        <span>
          Titulo Original:
        </span>
         
      </p>
      <p>
        <span>
          Genero:
        </span>
         
        Comedia
      </p>
      <p>
        <span>
          Pais de produccion:
        </span>
         
        Mexico
      </p>
      <p>
        <span>
          Lanzamiento:
        </span>
         
        11 agosto 2020
      </p>
      <p>
        <span>
          Likes:
        </span>
         
      </p>
      <p>
        <span>
          Calificacion:
        </span>
         
      </p>
      <p>
        <span>
          Numero de votos:
        </span>
         
      </p>
      <div
        className="sc-AxhUy bRWbEk"
      >
        <p
          style={
            Object {
              "color": "#ff6d00",
              "fontWeight": "bold",
            }
          }
        >
          Añadir a favoritos
        </p>
        <i
          className="fa"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Test in <Movie /> the film must be shown correctly 1`] = `
<div
  className="sc-AxjAm kpQkWH animate__animated animate__fadeIn"
>
  <div
    className="sc-AxirZ cTYuej"
  >
    <h2>
      Work It: Al ritmo de los sueños
    </h2>
    <p />
  </div>
  <div
    className="sc-AxiKw dqhJqz"
  >
    <img
      alt="No hay Imagen Disponible"
      src="https://images.unsplash.com/photo-1518676590629-3dcbd9c5a5c9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=80"
    />
  </div>
  <div
    className="sc-AxhCb fYZkjQ"
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <p />
      <p>
        <span>
          Titulo Original:
        </span>
         
      </p>
      <p>
        <span>
          Genero:
        </span>
         
        Comedia
      </p>
      <p>
        <span>
          Pais de produccion:
        </span>
         
        Mexico
      </p>
      <p>
        <span>
          Lanzamiento:
        </span>
         
        11 agosto 2020
      </p>
      <p>
        <span>
          Likes:
        </span>
         
      </p>
      <p>
        <span>
          Calificacion:
        </span>
         
      </p>
      <p>
        <span>
          Numero de votos:
        </span>
         
      </p>
    </div>
  </div>
</div>
`;
